File: services/vectorization-service/src/infrastructure/messaging/kafka_event_consumer.py
from kafka import KafkaConsumer
from typing import Callable, Dict, Any
import json
import os
import asyncio
from src.infrastructure.config.logger import logger


class KafkaEventConsumer:
    def __init__(self, group_id: str):
        kafka_broker = os.getenv("KAFKA_BROKER", "localhost:9092")
        self.consumer = KafkaConsumer(
            bootstrap_servers=[kafka_broker],
            group_id=group_id,
            value_deserializer=lambda m: json.loads(m.decode("utf-8")),
            auto_offset_reset="latest",
        )
        self.is_running = False

    async def start(self):
        if not self.is_running:
            self.is_running = True

    async def stop(self):
        if self.is_running:
            self.consumer.close()
            self.is_running = False

    async def subscribe(self, topic: str, handler: Callable[[Dict[str, Any]], None]):
        if not self.is_running:
            await self.start()

        self.consumer.subscribe([topic])

        async def consume_messages():
            for message in self.consumer:
                try:
                    await handler(message.value)
                except Exception as e:
                    logger.error("Error processing message from topic", topic=topic, error=str(e), exc_info=True)

        # Ejecutar en un loop separado
        asyncio.create_task(consume_messages())
        logger.info("Subscribed to topic", topic=topic)

File: services/audit-service/src/infrastructure/messaging/KafkaEventConsumer.ts
import { Kafka } from "kafkajs";
import type { IEventConsumer } from "../../application/ports/IEventConsumer.js";
import logger from '../config/logger.js';

export class KafkaEventConsumer implements IEventConsumer {
    private consumer: any;
    private isRunning: boolean = false;

    constructor() {
        const kafka = new Kafka({
            clientId: "audit-service",
            brokers: (process.env.KAFKA_BROKER || "localhost:9092").split(","),
        });

        this.consumer = kafka.consumer({ groupId: "audit-service-group" });
    }

    async start(): Promise<void> {
        if (this.isRunning) {
            return;
        }

        await this.consumer.connect();
        this.isRunning = true;
        logger.info({ message: 'Kafka consumer connected' });
    }

    async stop(): Promise<void> {
        if (!this.isRunning) {
            return;
        }

        await this.consumer.disconnect();
        this.isRunning = false;
        logger.info({ message: 'Kafka consumer disconnected' });
    }

    async subscribe(topic: string, handler: (message: any) => Promise<void>): Promise<void> {
        if (!this.isRunning) {
            await this.start();
        }

        await this.consumer.subscribe({ topic, fromBeginning: false });

        await this.consumer.run({
            eachMessage: async ({ message }: any) => {
                try {
                    const value = message.value?.toString();
                    if (value) {
                        const payload = JSON.parse(value);
                        await handler(payload);
                    }
                } catch (error) {
                    logger.error({ 
                        message: 'Error processing message from topic',
                        topic,
                        error: error instanceof Error ? error.message : String(error),
                        stack: error instanceof Error ? error.stack : undefined
                    });
                }
            },
        });

        logger.info({ 
            message: 'Subscribed to topic',
            topic 
        });
    }
}
